<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Phylogenetic Analysis of Genome-Influenced Traits • phyloAMR</title>
<script src="lightswitch.js"></script><script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Phylogenetic Analysis of Genome-Influenced Traits">
<meta name="description" content="Toolkit to perform phylogenetically-informed analysis of antibiotic resistance and other genome-influenced traits">
<meta property="og:description" content="Toolkit to perform phylogenetically-informed analysis of antibiotic resistance and other genome-influenced traits">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">phyloAMR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/ancestral_state_reconstruction_cluster_detection.html">Detection of phylogenetic clusters using phyloAMR</a></li>
    <li><a class="dropdown-item" href="articles/ancestral_state_reconstruction_of_a_trait.html">Ancestral state reconstruction of a trait</a></li>
    <li><a class="dropdown-item" href="articles/ancestral_state_reconstruction_phenotype_genotype_investigations.html">Phenotype and genotype association testing using ancestral state reconstruction</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/kylegontjes/phyloAMR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="phyloamr-an-r-package-to-perform-phylogenetic-analysis-of-antimicrobial-resistance">phyloAMR: an R package to perform PHYLOgenetic analysis of AntiMicrobial Resistance<a class="anchor" aria-label="anchor" href="#phyloamr-an-r-package-to-perform-phylogenetic-analysis-of-antimicrobial-resistance"></a>
</h1></div>
<p><a href="https://github.com/kylegontjes/phyloAMR/actions/workflows/ci.yml" class="external-link"><img src="https://github.com/kylegontjes/phyloAMR/actions/workflows/ci.yml/badge.svg" alt="R CMD Build"></a> <a href="https://www.repostatus.org/#active" class="external-link"><img src="https://www.repostatus.org/badges/latest/active.svg" alt="Project Status: Active - The project has reached a stable, usable state and is being actively developed."></a> <a href="https://cran.r-project.org/web/licenses/MIT" class="external-link"><img src="https://img.shields.io/badge/license-MIT-blue.svg" alt="License: MIT"></a> <a href="%5B%5Bhttps://example.com%5D(https://kylegontjes.github.io/phyloAMR/)%5D"><img src="https://img.shields.io/badge/Website-Visit-blue" alt="Visit Website"></a></p>
<p><strong>Description</strong></p>
<p>phyloAMR is an R package that characterizes the evolutionary history of genome-influenced traits and investigate phenotype-genotype associations. This package can be applied to genome-influenced traits, including antibiotic resistance, infection severity, source of infection, and genotypes.</p>
<p>Leveraging corHMM’s ancestral state reconstruction, this package characterizes a trait’s gain, loss, and continuation across a phylogeny. Inferred ancestral states can address numerous research questions:<br>
1. How often does an antibiotic-resistant bacterium emerge and spread across a healthcare network? 2. Are certain genotypes associated with the gain or loss of antibiotic resistance?<br>
3. Do compensatory or revertant mutations follow the acquisition of a genome-informed trait? 4. What patient characteristics are associated with the emergence and spread of antibiotic-resistant organisms?</p>
<p>Additional phylogenetically-informed algorithms, that do not use ancestral state reconstruction, have been developed: 1. <strong>nearest_neighbor_algorithm()</strong>: Identification of an isolate’s nearest neighbor on the phylogenetic tree. Can be used to identify an isolate’s nearest neighbor without a feature (i.e., susceptible neighbor of a resistant isolate). This method can be combined with nearest_neighbor_analysis() and nearest_neighbor_summary_statistic() to compare characteristics of the isolate and their nearest neighbor. This analysis framework is useful for identifying the influence of genetic mutations on bacterial phenotypes (i.e., acquisition of resistance mutation and a resistance measurement). 2. <strong>phyloaware_regression()</strong>: Phylogenetically-informed regression to test for associations between descriptive characteristics and the emergence and spread of a genome-influenced trait.</p>
<p><strong>How to install phyloAMR</strong></p>
<pre><code><span><span class="co"># Install devtools</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span>, dep <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Install corHMM &amp; phyloAMR from GitHub</span></span>
<span><span class="co">## corHMM - installed because the GitHub version is preferred</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"thej022214/corHMM"</span>, force <span class="op">=</span> <span class="cn">TRUE</span>, build_vignettes <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## phyloAMR</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"kylegontjes/phyloAMR"</span>, force <span class="op">=</span> <span class="cn">TRUE</span>, build_vignettes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
<p><strong>Questions that this tool can address</strong></p>
<table class="table">
<colgroup>
<col width="20%">
<col width="20%">
<col width="20%">
<col width="20%">
<col width="20%">
</colgroup>
<thead><tr class="header">
<th>Question</th>
<th>Method</th>
<th>Function(s)</th>
<th>Inputs</th>
<th>Output</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>What are the ancestral states for our trait?</td>
<td>Ancestral state reconstruction using corHMM</td>
<td>asr()</td>
<td>Dataframe with a trait and a phylogenetic tree</td>
<td>Ancestral reconstruction states</td>
</tr>
<tr class="even">
<td>How often is the trait gained and lost across the phylogeny?</td>
<td>Ancestral state reconstruction and transition statistics</td>
<td>asr() + asr_transition_analysis()</td>
<td>Ancestral reconstruction states from asr()</td>
<td>Descriptive statistics on trait gain, loss, and continuation</td>
</tr>
<tr class="odd">
<td>Do isolates belong to episodes of trait emergence or spread?</td>
<td>Ancestral state reconstruction and tree traversal of estimates</td>
<td>asr() + asr_cluster_detection()</td>
<td>Dataframe with genome-influenced trait and a phylogenetic tree</td>
<td>Ancestral reconstruction and calls for phylogenetic clusters and singletons</td>
</tr>
<tr class="even">
<td>How often does a genome-informed trait emerge and spread across a healthcare network?</td>
<td>Ancestral state reconstruction and tracing of estimates</td>
<td>asr() + asr_cluster_detection() + asr_cluster_analysis()</td>
<td>Dataframe with genome-influenced trait and a phylogenetic tree</td>
<td>Descriptive statistics on phylogenetic cluster calls and singletons</td>
</tr>
<tr class="odd">
<td>Genetic features associated with trait/lineage emergence/spread?</td>
<td>Ancestral state reconstruction of phenotype and genotype</td>
<td>asr() + synchronous_transitions() + synchronous_permutation_test()</td>
<td>Dataframe with genome-influenced trait + genotypes of interest and a phylogenetic tree</td>
<td>Two traits with synchronous episodes of gain or loss</td>
</tr>
<tr class="even">
<td>What genotypes are subsequently gained/lost upon acquisition of a trait (i.e., potential compensatory or revertant mutations)</td>
<td>Ancestral state reconstruction of phenotype and genotypes</td>
<td>asr() + downstream_transitions() + downstream_permutation_test()</td>
<td>Dataframe with genome-influenced trait + genotypes of interest and a phylogenetic tree</td>
<td>Genotypes classified as downstream mutations from a traits gain event</td>
</tr>
<tr class="odd">
<td>What descriptive characteristics are associated with phenotypic emergence and spread?</td>
<td>Association testing for characteristics and trait emergence (singletons) and spread (clusters)</td>
<td>phyloaware_regression()</td>
<td>Cluster calls from asr_cluster_detection() algorithm and a dataframe with characteristics of interest</td>
<td>Statistical association testing results for characteristics of interest</td>
</tr>
<tr class="even">
<td>Who is my isolate’s closest related strain (i.e., nearest neighbor)?</td>
<td>Evaluation of phylogenetic distance</td>
<td>nearest_neighbor_algorithm()</td>
<td>Dataframe with genome-influenced trait and a phylogenetic tree</td>
<td>An isolate’s nearest neighbor</td>
</tr>
</tbody>
</table>
<p><strong>Citation</strong></p>
<p>Kyle J Gontjes, Aryan Singh, Sarah E Sansom, James D Boyko, Stephen A Smith, Ebbing Lautenbach, Evan Snitkin, Phylogenetic Context of Antibiotic Resistance Provides Insights into the Dynamics of Resistance Emergence and Spread, The Journal of Infectious Diseases, 2025;, jiaf478, <a href="https://doi.org/10.1093/infdis/jiaf478" class="external-link uri">https://doi.org/10.1093/infdis/jiaf478</a></p>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/kylegontjes/phyloAMR/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/kylegontjes/phyloAMR/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing phyloAMR</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Kyle Gontjes <br><small class="roles"> Maintainer, author </small> <a href="https://orcid.org/0000-0001-8671-7990" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
<li>Aryan Singh <br><small class="roles"> Author </small>   </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Kyle Gontjes, Aryan Singh.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
