<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Phenotype and genotype association testing using ancestral state reconstruction • phyloAMR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Phenotype and genotype association testing using ancestral state reconstruction">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">phyloAMR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/ancestral_state_reconstruction_cluster_detection.html">ancestral_state_reconstruction_cluster_detection</a></li>
    <li><a class="dropdown-item" href="../articles/ancestral_state_reconstruction_of_a_trait.html">ancestral_state_reconstruction_of_a_trait</a></li>
    <li><a class="dropdown-item" href="../articles/ancestral_state_reconstruction_phenotype_genotype_investigations.html">Phenotype and genotype association testing using ancestral state reconstruction</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/kylegontjes/phyloAMR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Phenotype and genotype association testing using ancestral state reconstruction</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/kylegontjes/phyloAMR/blob/master/vignettes/ancestral_state_reconstruction_phenotype_genotype_investigations.Rmd" class="external-link"><code>vignettes/ancestral_state_reconstruction_phenotype_genotype_investigations.Rmd</code></a></small>
      <div class="d-none name"><code>ancestral_state_reconstruction_phenotype_genotype_investigations.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This tool’s main workhorse is the <strong>asr()</strong>, which
leverages corHMM’s ancestral state reconstruction algorithm to
characterize genome-influenced features’ gain, loss, and continuation
across a phylogenetic tree.</p>
<p>After performing joint or ancestral state reconstruction, the
resultant state predictions of ancestral nodes are parsed to generate a
parent-child dataframe. By traversing the phylogenetic tree from the
tips to the root, the episodes of trait gain, loss, and continuation are
added to the phylogenetic tree’s edge matrix.</p>
<p>This information can be leveraged to infer whether a genotype is
associated with a phenotype. We wrote two functions to assess this: 1.
synchronous_transitions(): Detection of synchronous genotype-phenotype
transitions 2. downstream_transitions(): Detection of downstream
genotype gain/loss events on phenotypic stretches</p>
</div>
<div class="section level2">
<h2 id="environment">Environment<a class="anchor" aria-label="anchor" href="#environment"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/kylegontjes/phyloAMR" class="external-link">phyloAMR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/emmanuelparadis/ape" class="external-link">ape</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'ape'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:dplyr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     where</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.amazon.com/Integration-Manipulation-Visualization-Phylogenetic-Computational-ebook/dp/B0B5NLZR1Z/" class="external-link">ggtree</a></span><span class="op">)</span> </span>
<span><span class="co">#&gt; ggtree v3.16.0 Learn more at https://yulab-smu.top/contribution-tree-data/</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Please cite:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Guangchuang Yu, David Smith, Huachen Zhu, Yi Guan, Tommy Tsan-Yuk Lam.</span></span>
<span><span class="co">#&gt; ggtree: an R package for visualization and annotation of phylogenetic</span></span>
<span><span class="co">#&gt; trees with their covariates and other associated data. Methods in</span></span>
<span><span class="co">#&gt; Ecology and Evolution. 2017, 8(1):28-36. doi:10.1111/2041-210X.12628</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'ggtree'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:ape':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     rotate</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="load-example-data">Load example data<a class="anchor" aria-label="anchor" href="#load-example-data"></a>
</h2>
<p>This tutorial will focus on the emergence and spread of colistin
non-susceptibility in a collection of 413 carbapenem-resistant
<em>Klebsiella pneumoniae</em> specimens collected across 12 California
long-term acute care hospitals.</p>
<p>We focus on the evolution and spread of non-susceptibility to
colistin, a last-resort antibiotic used to treat Gram-negative
bacteria.</p>
<div class="section level3">
<h3 id="dataframe">Dataframe<a class="anchor" aria-label="anchor" href="#dataframe"></a>
</h3>
<ol style="list-style-type: decimal">
<li>
<strong>tip_name_variable:</strong> variable with tip names</li>
<li>
<strong>Patient_ID:</strong> Identifiers for patients in this
study</li>
<li>
<strong>clades:</strong> what clade of epidemic lineage sequence
type 258 the isolate belongs to</li>
<li>
<strong>colistin_ns:</strong> colistin non-susceptibility. 1 =
non-susceptible. 0 = susceptible</li>
</ol>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu">phyloAMR</span><span class="fu">::</span><span class="va"><a href="../reference/df.html">df</a></span>  </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Total of 413 isolates"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Total of 413 isolates"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Number of patients: "</span>,<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">Patient_ID</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Number of patients: 338"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 413   4</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"View of dataframe for first 5 isolates"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "View of dataframe for first 5 isolates"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">df</span>,n <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt;         tip_name_var Patient_ID    clades colistin_ns</span></span>
<span><span class="co">#&gt; PCMP_H1      PCMP_H1          1 clade IIA           0</span></span>
<span><span class="co">#&gt; PCMP_H2      PCMP_H2          2 clade IIA           0</span></span>
<span><span class="co">#&gt; PCMP_H3      PCMP_H3          3 clade IIB           1</span></span>
<span><span class="co">#&gt; PCMP_H4      PCMP_H4          4 clade IIB           0</span></span>
<span><span class="co">#&gt; PCMP_H5      PCMP_H5          5   clade I           0</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Frequency of non-susceptibility to colistin: "</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Frequency of non-susceptibility to colistin: "</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">colistin_ns</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   0   1 </span></span>
<span><span class="co">#&gt; 277 136</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="phylogenetic-tree">Phylogenetic tree<a class="anchor" aria-label="anchor" href="#phylogenetic-tree"></a>
</h3>
<p>This maximum-likelihood phylogenetic tree was constructed on a
Gubbins recombination filtered alignment of single-nucleotide
polymorphisms using IQ-TREE</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tr</span> <span class="op">&lt;-</span> <span class="fu">phyloAMR</span><span class="fu">::</span><span class="va"><a href="../reference/tr.html">tr</a></span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html" class="external-link">ggtree</a></span><span class="op">(</span><span class="va">tr</span><span class="op">)</span></span></code></pre></div>
<p><img src="ancestral_state_reconstruction_phenotype_genotype_investigations_files/figure-html/unnamed-chunk-3-1.png" width="720"></p>
</div>
</div>
<div class="section level2">
<h2 id="step-0-1-visualization-of-the-trait-across-the-phylogenetic-tree">Step 0.1: Visualization of the trait across the phylogenetic
tree<a class="anchor" aria-label="anchor" href="#step-0-1-visualization-of-the-trait-across-the-phylogenetic-tree"></a>
</h2>
<p>Before performing ancestral state reconstruction, it is critical to
visualize the tip states of the trait on the phylogeny.</p>
<p>Our ancestral state reconstruction and clustering algorithm is most
powerful in settings where frequent emergence and spread of a trait
occurs.</p>
<p>Notice the clustering of Colistin non-susceptibility across the
phylogeny.</p>
<ul>
<li><p>Numerous emergence events at ancestral nodes and tips can be
visually inferred in this phylogeny</p></li>
<li><p>Ancestral emergence events had noticeable variations in size. For
instance, consider the large cluster of non-susceptibility in the above
clade</p></li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>`1` <span class="op">=</span> <span class="st">"black"</span>,`0`<span class="op">=</span><span class="st">"white"</span><span class="op">)</span></span>
<span><span class="va">feature_scale</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="va">feature_colors</span>,labels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Present"</span>,<span class="st">"Absent"</span><span class="op">)</span>,name<span class="op">=</span><span class="st">"Tip State"</span>, guide <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>nrow<span class="op">=</span><span class="fl">1</span>, title.position <span class="op">=</span> <span class="st">"top"</span>, label.position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p0</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/gheatmap.html" class="external-link">gheatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html" class="external-link">ggtree</a></span><span class="op">(</span><span class="va">tr</span><span class="op">)</span>,<span class="va">df</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">colistin_ns</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate_all.html" class="external-link">mutate_all</a></span><span class="op">(</span><span class="va">as.factor</span><span class="op">)</span>,colnames_position <span class="op">=</span> <span class="st">'top'</span>,width <span class="op">=</span> <span class="fl">.25</span>,low <span class="op">=</span> <span class="st">'white'</span>,high<span class="op">=</span><span class="st">'black'</span>,colnames_angle <span class="op">=</span> <span class="fl">90</span>,legend_title <span class="op">=</span> <span class="st">'Colistin non-susceptibility'</span>,hjust <span class="op">=</span> <span class="fl">0</span>,color <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="cn">NA</span>,<span class="fl">485</span><span class="op">)</span> <span class="op">+</span> <span class="va">feature_scale</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">y</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">y</span>, which will replace the existing scale.</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">fill</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span></span>
<span><span class="va">p0</span></span></code></pre></div>
<p><img src="ancestral_state_reconstruction_phenotype_genotype_investigations_files/figure-html/unnamed-chunk-4-1.png" width="720"></p>
</div>
<div class="section level2">
<h2 id="step-0-2-visualization-of-the-genotypes-across-the-phylogenetic-tree">Step 0.2: Visualization of the genotypes across the phylogenetic
tree<a class="anchor" aria-label="anchor" href="#step-0-2-visualization-of-the-genotypes-across-the-phylogenetic-tree"></a>
</h2>
<p>We provide genotypes associated with colistin non-susceptibility.
These can be leveraged to demonstrate genotype-phenotype algorithms
developed in this package.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genotype_mat</span> <span class="op">&lt;-</span> <span class="fu">phyloAMR</span><span class="fu">::</span><span class="va"><a href="../reference/genotype_mat.html">genotype_mat</a></span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Total of "</span>,<span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">genotype_mat</span><span class="op">)</span>,<span class="st">" genotypes"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Total of 76 genotypes"</span></span>
<span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Genotypes present in this matrix: "</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Genotypes present in this matrix: "</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">genotype_mat</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "Coding SNP at 3359989 &gt; T functional=NULL_NULL_NULL locus_tag=KPNIH1_RS30220 Strand Information: KPNIH1_RS30220=-;T|stop_gained|HIGH|KPNIH1_RS30220|c.141G&gt;A|p.Trp47*|141/144|47/47|mgrB|PhoP/PhoQ regulator MgrB;_68746"                                                                                                 </span></span>
<span><span class="co">#&gt;  [2] "Coding SNP at 3360003 &gt; A functional=NULL_NULL_NULL locus_tag=KPNIH1_RS30220 Strand Information: KPNIH1_RS30220=-;A|stop_gained|HIGH|KPNIH1_RS30220|c.127A&gt;T|p.Lys43*|127/144|43/47|mgrB|PhoP/PhoQ regulator MgrB;_68748"                                                                                                 </span></span>
<span><span class="co">#&gt;  [3] "Coding SNP at 3360042 &gt; A functional=NULL_NULL_NULL locus_tag=KPNIH1_RS30220 Strand Information: KPNIH1_RS30220=-;A|stop_gained|HIGH|KPNIH1_RS30220|c.88C&gt;T|p.Gln30*|88/144|30/47|mgrB|PhoP/PhoQ regulator MgrB;_68751"                                                                                                   </span></span>
<span><span class="co">#&gt;  [4] "Coding SNP at 3360066 &gt; A functional=NULL_NULL_NULL locus_tag=KPNIH1_RS30220 Strand Information: KPNIH1_RS30220=-;A|stop_gained|HIGH|KPNIH1_RS30220|c.64C&gt;T|p.Gln22*|64/144|22/47|mgrB|PhoP/PhoQ regulator MgrB;_68754"                                                                                                   </span></span>
<span><span class="co">#&gt;  [5] "Coding SNP at 3360123 &gt; A functional=NULL_NULL_NULL locus_tag=KPNIH1_RS30220 Strand Information: KPNIH1_RS30220=-;A|stop_gained|HIGH|KPNIH1_RS30220|c.7A&gt;T|p.Lys3*|7/144|3/47|mgrB|PhoP/PhoQ regulator MgrB;_68758"                                                                                                       </span></span>
<span><span class="co">#&gt;  [6] "Coding Indel at 3360097 &gt; GA functional=NULL_NULL_NULL locus_tag=KPNIH1_RS30220 Strand Information: KPNIH1_RS30220=-;GA|frameshift_variant|HIGH|KPNIH1_RS30220|c.32dupT|p.Ile12fs|32/144|11/47|mgrB|PhoP/PhoQ regulator MgrB;_1973"                                                                                       </span></span>
<span><span class="co">#&gt;  [7] "Coding Indel at 3360098 &gt; A functional=NULL_NULL_NULL locus_tag=KPNIH1_RS30220 Strand Information: KPNIH1_RS30220=-;A|frameshift_variant|HIGH|KPNIH1_RS30220|c.24_31delACTGATAG|p.Leu8fs|31/144|8/47|mgrB|PhoP/PhoQ regulator MgrB;_1974"                                                                                 </span></span>
<span><span class="co">#&gt;  [8] "3360144_3360148|KPNIH1_RS30220-KPNIH1_RS17035|-.+"                                                                                                                                                                                                                                                                        </span></span>
<span><span class="co">#&gt;  [9] "3360057_3360060|KPNIH1_RS30220|-"                                                                                                                                                                                                                                                                                         </span></span>
<span><span class="co">#&gt; [10] "3359997_3360004|KPNIH1_RS30220|-"                                                                                                                                                                                                                                                                                         </span></span>
<span><span class="co">#&gt; [11] "3360056_3360060|KPNIH1_RS30220|-"                                                                                                                                                                                                                                                                                         </span></span>
<span><span class="co">#&gt; [12] "3360032_3360036|KPNIH1_RS30220|-"                                                                                                                                                                                                                                                                                         </span></span>
<span><span class="co">#&gt; [13] "Coding SNP at 1693296 &gt; C functional=NULL_NULL_NULL locus_tag=KPNIH1_RS08505 Strand Information: KPNIH1_RS08505=-;C|missense_variant|MODERATE|KPNIH1_RS08505|c.245T&gt;G|p.Leu82Arg|245/1098|82/365|pmrB|two-component system sensor histidine kinase PmrB;_32525"                                                           </span></span>
<span><span class="co">#&gt; [14] "Coding SNP at 3058674 &gt; T functional=NULL_NULL_NULL locus_tag=KPNIH1_RS15525 Strand Information: KPNIH1_RS15525=+;T|missense_variant|MODERATE|KPNIH1_RS15525|c.29A&gt;T|p.Gln10Leu|29/1062|10/353|ccrB|HAMP domain-containing histidine kinase CrrB;_62072"                                                                  </span></span>
<span><span class="co">#&gt; [15] "Coding SNP at 3058925 &gt; A functional=NULL_NULL_NULL locus_tag=KPNIH1_RS15525 Strand Information: KPNIH1_RS15525=+;A|missense_variant|MODERATE|KPNIH1_RS15525|c.280C&gt;A|p.Leu94Met|280/1062|94/353|ccrB|HAMP domain-containing histidine kinase CrrB;_62078"                                                                </span></span>
<span><span class="co">#&gt; [16] "Coding SNP at 3059096 &gt; A,T functional=NULL_NULL_NULL locus_tag=KPNIH1_RS15525 Strand Information: KPNIH1_RS15525=+;T|missense_variant|MODERATE|KPNIH1_RS15525|c.451C&gt;T|p.Pro151Ser|451/1062|151/353|ccrB|ccrB;A|missense_variant|MODERATE|KPNIH1_RS15525|c.451C&gt;A|p.Pro151Thr|451/1062|151/353|ccrB|ccrB;_62079"         </span></span>
<span><span class="co">#&gt; [17] "Coding SNP at 3059096 &gt; A,T functional=NULL_NULL_NULL locus_tag=KPNIH1_RS15525 Strand Information: KPNIH1_RS15525=+;T|missense_variant|MODERATE|KPNIH1_RS15525|c.451C&gt;T|p.Pro151Ser|451/1062|151/353|ccrB|ccrB;A|missense_variant|MODERATE|KPNIH1_RS15525|c.451C&gt;A|p.Pro151Thr|451/1062|151/353|ccrB|ccrB;_62080"         </span></span>
<span><span class="co">#&gt; [18] "Coding SNP at 3059193 &gt; T functional=NULL_NULL_NULL locus_tag=KPNIH1_RS15525 Strand Information: KPNIH1_RS15525=+;T|missense_variant|MODERATE|KPNIH1_RS15525|c.548G&gt;T|p.Gly183Val|548/1062|183/353|ccrB|HAMP domain-containing histidine kinase CrrB;_62085"                                                              </span></span>
<span><span class="co">#&gt; [19] "Coding SNP at 3360039 &gt; T functional=NULL_NULL_NULL locus_tag=KPNIH1_RS30220 Strand Information: KPNIH1_RS30220=-;T|missense_variant|MODERATE|KPNIH1_RS30220|c.91G&gt;A|p.Asp31Asn|91/144|31/47|mgrB|PhoP/PhoQ regulator MgrB;_68750"                                                                                        </span></span>
<span><span class="co">#&gt; [20] "Coding SNP at 2080519 &gt; T functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10370 Strand Information: KPNIH1_RS10370=-;T|missense_variant|MODERATE|KPNIH1_RS10370|c.731C&gt;A|p.Thr244Asn|731/1467|244/488|phoQ|two-component system sensor histidine kinase PhoQ;_38555"                                                         </span></span>
<span><span class="co">#&gt; [21] "Coding SNP at 1693068 &gt; T functional=NULL_NULL_NULL locus_tag=KPNIH1_RS08505 Strand Information: KPNIH1_RS08505=-;T|missense_variant|MODERATE|KPNIH1_RS08505|c.473C&gt;A|p.Pro158Gln|473/1098|158/365|pmrB|two-component system sensor histidine kinase PmrB;_32517"                                                         </span></span>
<span><span class="co">#&gt; [22] "Coding SNP at 1694059 &gt; T functional=NULL_NULL_NULL locus_tag=KPNIH1_RS08510 Strand Information: KPNIH1_RS08510=-;T|missense_variant|MODERATE|KPNIH1_RS08510|c.157G&gt;A|p.Gly53Ser|157/672|53/223|pmrA|two-component system response regulator PmrA;_32532"                                                                 </span></span>
<span><span class="co">#&gt; [23] "Coding SNP at 3360008 &gt; T functional=NULL_NULL_NULL locus_tag=KPNIH1_RS30220 Strand Information: KPNIH1_RS30220=-;T|missense_variant|MODERATE|KPNIH1_RS30220|c.122T&gt;A|p.Ile41Asn|122/144|41/47|mgrB|PhoP/PhoQ regulator MgrB;_68749"                                                                                      </span></span>
<span><span class="co">#&gt; [24] "Coding SNP at 3360056 &gt; G functional=NULL_NULL_NULL locus_tag=KPNIH1_RS30220 Strand Information: KPNIH1_RS30220=-;G|missense_variant|MODERATE|KPNIH1_RS30220|c.74A&gt;C|p.Asn25Thr|74/144|25/47|mgrB|PhoP/PhoQ regulator MgrB;_68752"                                                                                        </span></span>
<span><span class="co">#&gt; [25] "Non-Coding SNP at 3360138 &gt; A functional=NULL_NULL_NULL locus_tag=KPNIH1_RS30220-KPNIH1_RS17035 Strand Information: KPNIH1_RS30220=-/KPNIH1_RS17035=+/;A|intergenic_region|MODIFIER|KPNIH1_RS30220-KPNIH1_RS17035|n.3360138C&gt;A||||mgrB,null or hypothetical protein|PhoP/PhoQ regulator MgrB,hypothetical protein,;_68759"</span></span>
<span><span class="co">#&gt; [26] "Non-Coding SNP at 3360139 &gt; A functional=NULL_NULL_NULL locus_tag=KPNIH1_RS30220-KPNIH1_RS17035 Strand Information: KPNIH1_RS30220=-/KPNIH1_RS17035=+/;A|intergenic_region|MODIFIER|KPNIH1_RS30220-KPNIH1_RS17035|n.3360139C&gt;A||||mgrB,null or hypothetical protein|PhoP/PhoQ regulator MgrB,hypothetical protein,;_68760"</span></span>
<span><span class="co">#&gt; [27] "Non-Coding SNP at 3360169 &gt; T functional=NULL_NULL_NULL locus_tag=KPNIH1_RS30220-KPNIH1_RS17035 Strand Information: KPNIH1_RS30220=-/KPNIH1_RS17035=+/;T|intergenic_region|MODIFIER|KPNIH1_RS30220-KPNIH1_RS17035|n.3360169A&gt;T||||mgrB,null or hypothetical protein|PhoP/PhoQ regulator MgrB,hypothetical protein,;_68761"</span></span>
<span><span class="co">#&gt; [28] "Non-Coding SNP at 3360177 &gt; T functional=NULL_NULL_NULL locus_tag=KPNIH1_RS30220-KPNIH1_RS17035 Strand Information: KPNIH1_RS30220=-/KPNIH1_RS17035=+/;T|intergenic_region|MODIFIER|KPNIH1_RS30220-KPNIH1_RS17035|n.3360177C&gt;T||||mgrB,null or hypothetical protein|PhoP/PhoQ regulator MgrB,hypothetical protein,;_68762"</span></span>
<span><span class="co">#&gt; [29] "Coding SNP at 2081763 &gt; A functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10375 Strand Information: KPNIH1_RS10375=-;A|missense_variant|MODERATE|KPNIH1_RS10375|c.158G&gt;T|p.Gly53Val|158/672|53/223|phoP|two-component system response regulator PhoP;_38580"                                                                 </span></span>
<span><span class="co">#&gt; [30] "Coding SNP at 2080327 &gt; T functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10370 Strand Information: KPNIH1_RS10370=-;T|missense_variant|MODERATE|KPNIH1_RS10370|c.923T&gt;A|p.Leu308Gln|923/1467|308/488|phoQ|two-component system sensor histidine kinase PhoQ;_38551"                                                         </span></span>
<span><span class="co">#&gt; [31] "Coding SNP at 2080480 &gt; T functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10370 Strand Information: KPNIH1_RS10370=-;T|missense_variant|MODERATE|KPNIH1_RS10370|c.770T&gt;A|p.Leu257Gln|770/1467|257/488|phoQ|two-component system sensor histidine kinase PhoQ;_38553"                                                         </span></span>
<span><span class="co">#&gt; [32] "Coding SNP at 2080508 &gt; G functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10370 Strand Information: KPNIH1_RS10370=-;G|missense_variant|MODERATE|KPNIH1_RS10370|c.742A&gt;C|p.Thr248Pro|742/1467|248/488|phoQ|two-component system sensor histidine kinase PhoQ;_38554"                                                         </span></span>
<span><span class="co">#&gt; [33] "Coding SNP at 2080936 &gt; T functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10370 Strand Information: KPNIH1_RS10370=-;T|missense_variant|MODERATE|KPNIH1_RS10370|c.314T&gt;A|p.Leu105Gln|314/1467|105/488|phoQ|two-component system sensor histidine kinase PhoQ;_38561"                                                         </span></span>
<span><span class="co">#&gt; [34] "Coding SNP at 2081065 &gt; G functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10370 Strand Information: KPNIH1_RS10370=-;G|missense_variant|MODERATE|KPNIH1_RS10370|c.185T&gt;C|p.Leu62Pro|185/1467|62/488|phoQ|two-component system sensor histidine kinase PhoQ;_38564"                                                           </span></span>
<span><span class="co">#&gt; [35] "Coding SNP at 1692756 &gt; G functional=NULL_NULL_NULL locus_tag=KPNIH1_RS08505 Strand Information: KPNIH1_RS08505=-;G|missense_variant|MODERATE|KPNIH1_RS08505|c.785T&gt;C|p.Leu262Pro|785/1098|262/365|pmrB|two-component system sensor histidine kinase PmrB;_32509"                                                         </span></span>
<span><span class="co">#&gt; [36] "Coding SNP at 1692934 &gt; G functional=NULL_NULL_NULL locus_tag=KPNIH1_RS08505 Strand Information: KPNIH1_RS08505=-;G|missense_variant|MODERATE|KPNIH1_RS08505|c.607T&gt;C|p.Ser203Pro|607/1098|203/365|pmrB|two-component system sensor histidine kinase PmrB;_32516"                                                         </span></span>
<span><span class="co">#&gt; [37] "Coding SNP at 3059243 &gt; C functional=NULL_NULL_NULL locus_tag=KPNIH1_RS15525 Strand Information: KPNIH1_RS15525=+;C|missense_variant|MODERATE|KPNIH1_RS15525|c.598G&gt;C|p.Ala200Pro|598/1062|200/353|ccrB|HAMP domain-containing histidine kinase CrrB;_62089"                                                              </span></span>
<span><span class="co">#&gt; [38] "3059600_3059605|KPNIH1_RS15525|+"                                                                                                                                                                                                                                                                                         </span></span>
<span><span class="co">#&gt; [39] "Coding SNP at 3360069 &gt; G functional=NULL_NULL_NULL locus_tag=KPNIH1_RS30220 Strand Information: KPNIH1_RS30220=-;G|missense_variant|MODERATE|KPNIH1_RS30220|c.61A&gt;C|p.Thr21Pro|61/144|21/47|mgrB|PhoP/PhoQ regulator MgrB;_68756"                                                                                        </span></span>
<span><span class="co">#&gt; [40] "Coding SNP at 2081316 &gt; G functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10375 Strand Information: KPNIH1_RS10375=-;G|missense_variant|MODERATE|KPNIH1_RS10375|c.605A&gt;C|p.Gln202Pro|605/672|202/223|phoP|two-component system response regulator PhoP;_38572"                                                               </span></span>
<span><span class="co">#&gt; [41] "Coding SNP at 2081356 &gt; G functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10375 Strand Information: KPNIH1_RS10375=-;G|missense_variant|MODERATE|KPNIH1_RS10375|c.565A&gt;C|p.Thr189Pro|565/672|189/223|phoP|two-component system response regulator PhoP;_38573"                                                               </span></span>
<span><span class="co">#&gt; [42] "Coding SNP at 2081357 &gt; T functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10375 Strand Information: KPNIH1_RS10375=-;T|missense_variant|MODERATE|KPNIH1_RS10375|c.564C&gt;A|p.His188Gln|564/672|188/223|phoP|two-component system response regulator PhoP;_38574"                                                               </span></span>
<span><span class="co">#&gt; [43] "Coding SNP at 2081358 &gt; C functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10375 Strand Information: KPNIH1_RS10375=-;C|missense_variant|MODERATE|KPNIH1_RS10375|c.563A&gt;G|p.His188Arg|563/672|188/223|phoP|two-component system response regulator PhoP;_38575"                                                               </span></span>
<span><span class="co">#&gt; [44] "Coding SNP at 2081367 &gt; A functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10375 Strand Information: KPNIH1_RS10375=-;A|missense_variant|MODERATE|KPNIH1_RS10375|c.554G&gt;T|p.Arg185Leu|554/672|185/223|phoP|two-component system response regulator PhoP;_38576"                                                               </span></span>
<span><span class="co">#&gt; [45] "Coding SNP at 2081589 &gt; T functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10375 Strand Information: KPNIH1_RS10375=-;T|missense_variant|MODERATE|KPNIH1_RS10375|c.332G&gt;A|p.Arg111His|332/672|111/223|phoP|two-component system response regulator PhoP;_38578"                                                               </span></span>
<span><span class="co">#&gt; [46] "Coding SNP at 2081868 &gt; G functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10375 Strand Information: KPNIH1_RS10375=-;G|missense_variant|MODERATE|KPNIH1_RS10375|c.53T&gt;C|p.Val18Ala|53/672|18/223|phoP|two-component system response regulator PhoP;_38581"                                                                   </span></span>
<span><span class="co">#&gt; [47] "Coding Indel at 2081355 &gt; G functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10375 Strand Information: KPNIH1_RS10375=-;G|frameshift_variant|HIGH|KPNIH1_RS10375|c.565delA|p.Thr189fs|565/672|189/223|phoP|two-component system response regulator PhoP;_1050"                                                                </span></span>
<span><span class="co">#&gt; [48] "Coding Indel at 2081358 &gt; TC functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10375 Strand Information: KPNIH1_RS10375=-;TC|frameshift_variant|HIGH|KPNIH1_RS10375|c.562_563insG|p.His188fs|562/672|188/223|phoP|two-component system response regulator PhoP;_1051"                                                          </span></span>
<span><span class="co">#&gt; [49] "Coding SNP at 2079808 &gt; A functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10370 Strand Information: KPNIH1_RS10370=-;A|missense_variant|MODERATE|KPNIH1_RS10370|c.1442A&gt;T|p.Gln481Leu|1442/1467|481/488|phoQ|two-component system sensor histidine kinase PhoQ;_38535"                                                       </span></span>
<span><span class="co">#&gt; [50] "Coding SNP at 2079899 &gt; T functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10370 Strand Information: KPNIH1_RS10370=-;T|missense_variant|MODERATE|KPNIH1_RS10370|c.1351G&gt;A|p.Ala451Thr|1351/1467|451/488|phoQ|two-component system sensor histidine kinase PhoQ;_38537"                                                       </span></span>
<span><span class="co">#&gt; [51] "Coding SNP at 2079923 &gt; T functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10370 Strand Information: KPNIH1_RS10370=-;T|missense_variant|MODERATE|KPNIH1_RS10370|c.1327G&gt;A|p.Gly443Ser|1327/1467|443/488|phoQ|two-component system sensor histidine kinase PhoQ;_38539"                                                       </span></span>
<span><span class="co">#&gt; [52] "Coding SNP at 2079952 &gt; A functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10370 Strand Information: KPNIH1_RS10370=-;A|missense_variant|MODERATE|KPNIH1_RS10370|c.1298G&gt;T|p.Arg433Leu|1298/1467|433/488|phoQ|two-component system sensor histidine kinase PhoQ;_38540"                                                       </span></span>
<span><span class="co">#&gt; [53] "Coding SNP at 2080000 &gt; A functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10370 Strand Information: KPNIH1_RS10370=-;A|missense_variant|MODERATE|KPNIH1_RS10370|c.1250A&gt;T|p.Asp417Val|1250/1467|417/488|phoQ|two-component system sensor histidine kinase PhoQ;_38541"                                                       </span></span>
<span><span class="co">#&gt; [54] "Coding SNP at 2080286 &gt; T functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10370 Strand Information: KPNIH1_RS10370=-;T|missense_variant|MODERATE|KPNIH1_RS10370|c.964C&gt;A|p.Leu322Met|964/1467|322/488|phoQ|two-component system sensor histidine kinase PhoQ;_38550"                                                         </span></span>
<span><span class="co">#&gt; [55] "Coding SNP at 2080883 &gt; T functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10370 Strand Information: KPNIH1_RS10370=-;T|missense_variant|MODERATE|KPNIH1_RS10370|c.367G&gt;A|p.Glu123Lys|367/1467|123/488|phoQ|two-component system sensor histidine kinase PhoQ;_38558"                                                         </span></span>
<span><span class="co">#&gt; [56] "Coding SNP at 2081144 &gt; T functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10370 Strand Information: KPNIH1_RS10370=-;T|missense_variant|MODERATE|KPNIH1_RS10370|c.106G&gt;A|p.Ala36Thr|106/1467|36/488|phoQ|two-component system sensor histidine kinase PhoQ;_38567"                                                           </span></span>
<span><span class="co">#&gt; [57] "Coding Indel at 2080219 &gt; AG functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10370 Strand Information: KPNIH1_RS10370=-;AG|frameshift_variant|HIGH|KPNIH1_RS10370|c.1030dupC|p.Leu344fs|1030/1467|344/488|phoQ|two-component system sensor histidine kinase PhoQ;_1048"                                                      </span></span>
<span><span class="co">#&gt; [58] "Coding Indel at 2080486 &gt; T functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10370 Strand Information: KPNIH1_RS10370=-;T|disruptive_inframe_deletion|MODERATE|KPNIH1_RS10370|c.743_763delCCCGGCTGGTCAGCAACCTCA|p.Thr248_Leu254del|763/1467|248/488|phoQ|two-component system sensor histidine kinase PhoQ;_1049"             </span></span>
<span><span class="co">#&gt; [59] "Coding Indel at 3059657 &gt; A functional=NULL_NULL_NULL locus_tag=KPNIH1_RS15525 Strand Information: KPNIH1_RS15525=+;A|frameshift_variant|HIGH|KPNIH1_RS15525|c.1013_1016delGTTT|p.Ser338fs|1013/1062|338/353|ccrB|HAMP domain-containing histidine kinase CrrB;_1761"                                                     </span></span>
<span><span class="co">#&gt; [60] "Coding SNP at 3359999 &gt; C functional=NULL_NULL_NULL locus_tag=KPNIH1_RS30220 Strand Information: KPNIH1_RS30220=-;C|missense_variant|MODERATE|KPNIH1_RS30220|c.131T&gt;G|p.Phe44Cys|131/144|44/47|mgrB|PhoP/PhoQ regulator MgrB;_68747"                                                                                      </span></span>
<span><span class="co">#&gt; [61] "Coding SNP at 3360059 &gt; G functional=NULL_NULL_NULL locus_tag=KPNIH1_RS30220 Strand Information: KPNIH1_RS30220=-;G|missense_variant|MODERATE|KPNIH1_RS30220|c.71T&gt;C|p.Leu24Pro|71/144|24/47|mgrB|PhoP/PhoQ regulator MgrB;_68753"                                                                                        </span></span>
<span><span class="co">#&gt; [62] "Coding SNP at 3360074 &gt; T functional=NULL_NULL_NULL locus_tag=KPNIH1_RS30220 Strand Information: KPNIH1_RS30220=-;T|missense_variant|MODERATE|KPNIH1_RS30220|c.56T&gt;A|p.Leu19Gln|56/144|19/47|mgrB|PhoP/PhoQ regulator MgrB;_68757"                                                                                        </span></span>
<span><span class="co">#&gt; [63] "Coding SNP at 2081682 &gt; C functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10375 Strand Information: KPNIH1_RS10375=-;C|missense_variant|MODERATE|KPNIH1_RS10375|c.239C&gt;G|p.Ala80Gly|239/672|80/223|phoP|two-component system response regulator PhoP;_38579"                                                                 </span></span>
<span><span class="co">#&gt; [64] "Coding SNP at 2079865 &gt; C functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10370 Strand Information: KPNIH1_RS10370=-;C|missense_variant|MODERATE|KPNIH1_RS10370|c.1385T&gt;G|p.Ile462Ser|1385/1467|462/488|phoQ|two-component system sensor histidine kinase PhoQ;_38536"                                                       </span></span>
<span><span class="co">#&gt; [65] "Coding SNP at 2080156 &gt; T functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10370 Strand Information: KPNIH1_RS10370=-;T|missense_variant|MODERATE|KPNIH1_RS10370|c.1094T&gt;A|p.Leu365His|1094/1467|365/488|phoQ|two-component system sensor histidine kinase PhoQ;_38543"                                                       </span></span>
<span><span class="co">#&gt; [66] "Coding SNP at 2080222 &gt; A functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10370 Strand Information: KPNIH1_RS10370=-;A|missense_variant|MODERATE|KPNIH1_RS10370|c.1028C&gt;T|p.Pro343Leu|1028/1467|343/488|phoQ|two-component system sensor histidine kinase PhoQ;_38545"                                                       </span></span>
<span><span class="co">#&gt; [67] "Coding SNP at 2080256 &gt; G functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10370 Strand Information: KPNIH1_RS10370=-;G|missense_variant|MODERATE|KPNIH1_RS10370|c.994A&gt;C|p.Thr332Pro|994/1467|332/488|phoQ|two-component system sensor histidine kinase PhoQ;_38549"                                                         </span></span>
<span><span class="co">#&gt; [68] "Coding SNP at 2081101 &gt; A functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10370 Strand Information: KPNIH1_RS10370=-;A|missense_variant|MODERATE|KPNIH1_RS10370|c.149G&gt;T|p.Arg50Leu|149/1467|50/488|phoQ|two-component system sensor histidine kinase PhoQ;_38565"                                                           </span></span>
<span><span class="co">#&gt; [69] "Coding SNP at 2081153 &gt; T functional=NULL_NULL_NULL locus_tag=KPNIH1_RS10370 Strand Information: KPNIH1_RS10370=-;T|missense_variant|MODERATE|KPNIH1_RS10370|c.97G&gt;A|p.Gly33Ser|97/1467|33/488|phoQ|two-component system sensor histidine kinase PhoQ;_38568"                                                             </span></span>
<span><span class="co">#&gt; [70] "2080368_2080380|KPNIH1_RS10370|-"                                                                                                                                                                                                                                                                                         </span></span>
<span><span class="co">#&gt; [71] "Coding SNP at 1693257 &gt; T functional=NULL_NULL_NULL locus_tag=KPNIH1_RS08505 Strand Information: KPNIH1_RS08505=-;T|missense_variant|MODERATE|KPNIH1_RS08505|c.284C&gt;A|p.Pro95Gln|284/1098|95/365|pmrB|two-component system sensor histidine kinase PmrB;_32522"                                                           </span></span>
<span><span class="co">#&gt; [72] "1692837_1692844|KPNIH1_RS08505|-"                                                                                                                                                                                                                                                                                         </span></span>
<span><span class="co">#&gt; [73] "Coding SNP at 1152157 &gt; A functional=NULL_NULL_NULL locus_tag=KPNIH1_RS05725 Strand Information: KPNIH1_RS05725=-;A|missense_variant|MODERATE|KPNIH1_RS05725|c.215G&gt;T|p.Ser72Ile|215/288|72/95|null or hypothetical protein|PTS sugar transporter subunit IIB;_22458"                                                     </span></span>
<span><span class="co">#&gt; [74] "1152365_1152368|KPNIH1_RS05725|-"                                                                                                                                                                                                                                                                                         </span></span>
<span><span class="co">#&gt; [75] "Coding SNP at 4653807 &gt; C functional=NULL_NULL_NULL locus_tag=KPNIH1_RS23500 Strand Information: KPNIH1_RS23500=+;C|missense_variant|MODERATE|KPNIH1_RS23500|c.539T&gt;C|p.Val180Ala|539/1350|180/449|qseC|two-component system sensor histidine kinase QseC;_92679"                                                         </span></span>
<span><span class="co">#&gt; [76] "Coding Indel at 4653290 &gt; A functional=NULL_NULL_NULL locus_tag=KPNIH1_RS23500 Strand Information: KPNIH1_RS23500=+;A|conservative_inframe_deletion|MODERATE|KPNIH1_RS23500|c.34_39delCGCCTG|p.Arg12_Leu13del|34/1350|12/449|qseC|two-component system sensor histidine kinase QseC;_2579"</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Visualization of the genotypes"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Visualization of the genotypes"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/gheatmap.html" class="external-link">gheatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html" class="external-link">ggtree</a></span><span class="op">(</span><span class="va">tr</span><span class="op">)</span>,<span class="va">genotype_mat</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate_all.html" class="external-link">mutate_all</a></span><span class="op">(</span><span class="va">as.factor</span><span class="op">)</span>,width <span class="op">=</span> <span class="fl">0.8</span>,colnames <span class="op">=</span> <span class="cn">F</span><span class="op">)</span> <span class="op">+</span> <span class="va">feature_scale</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">y</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">y</span>, which will replace the existing scale.</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">fill</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span></span></code></pre></div>
<p><img src="ancestral_state_reconstruction_phenotype_genotype_investigations_files/figure-html/unnamed-chunk-5-1.png" width="960"></p>
</div>
<div class="section level2">
<h2 id="step-1-run-ancestral-state-reconstruction">Step 1: Run ancestral state reconstruction<a class="anchor" aria-label="anchor" href="#step-1-run-ancestral-state-reconstruction"></a>
</h2>
<ul>
<li><p>The workhorse function, asr(), is used to perform ancestral state
reconstruction. This wrapper function implements ancestral state
reconstruction with a single rate category using the corHMM R package:
<a href="https://github.com/thej022214/corHMM" class="external-link uri">https://github.com/thej022214/corHMM</a></p></li>
<li><p>Using inferred ancestral states and tip-based data, edges on the
phylogenetic tree were evaluated to determine episodes where the trait
continued (i.e., susceptible -&gt; susceptible or non-susceptible -&gt;
non-susceptible), was gained (i.e., susceptible -&gt; non-susceptible),
or was lost (i.e., non-susceptible -&gt; susceptible). This edge matrix
can be leveraged for numerous applications, including characterizing the
frequency of trait transitions across a phylogeny and investigating
phenotype-genotype associations.</p></li>
<li>
<p>The following parameters exist for this function</p>
<ul>
<li><p><strong>df:</strong> Dataframe with tip name variable (e.g.,
tip_name_variable) and trait variable (e.g., colistin_ns)</p></li>
<li><p><strong>tr:</strong> Phylogenetic tree object of class
phylo</p></li>
<li>
<p><strong>model:</strong> This approach permits the use of either
the equal rates (ER) or the all rates differ (ARD) transition
matrices.</p>
<ul>
<li><p>Equal rates: Assumes equal transition rates for trait gain (e.g.,
trait absence -&gt; presence) or loss (e.g., trait gain -&gt;
absence)</p></li>
<li><p>All rates differ: Assumes different transition rates for trait
gain and loss</p></li>
</ul>
</li>
<li>
<p>node_states: Whether to perform ‘joint’ or ‘marginal’ ancestral
state reconstruction</p>
<ul>
<li>From our experience, we recommend using joint ancestral state
reconstruction.</li>
</ul>
</li>
</ul>
</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">asr_obj</span> <span class="op">&lt;-</span> <span class="fu">phyloAMR</span><span class="fu">::</span><span class="fu"><a href="../reference/asr.html">asr</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">df</span>,tr <span class="op">=</span> <span class="va">tr</span>,tip_name_variable <span class="op">=</span> <span class="st">"tip_name_var"</span>,trait <span class="op">=</span> <span class="st">"colistin_ns"</span>,model <span class="op">=</span> <span class="st">"ER"</span>,node_states <span class="op">=</span> <span class="st">"joint"</span><span class="op">)</span></span>
<span><span class="co">#&gt; You specified 'fixed.nodes=FALSE' but included a phy object with node labels. These node labels have been removed.</span></span>
<span><span class="co">#&gt; State distribution in data:</span></span>
<span><span class="co">#&gt; States:  0   1   </span></span>
<span><span class="co">#&gt; Counts:  277 136 </span></span>
<span><span class="co">#&gt; Beginning thorough optimization search -- performing 0 random restarts </span></span>
<span><span class="co">#&gt; Finished. Inferring ancestral states using joint reconstruction.</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Output names: "</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Output names: "</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">asr_obj</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "corHMM_output"           "corHMM_model_statistics"</span></span>
<span><span class="co">#&gt; [3] "parent_child_df"         "node_states"</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"corHMM_out: output from ancestral state reconstruction algorithm hosted in the R package corHMM"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "corHMM_out: output from ancestral state reconstruction algorithm hosted in the R package corHMM"</span></span>
<span><span class="va">asr_obj</span><span class="op">$</span><span class="va">corHMM_out</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Fit</span></span>
<span><span class="co">#&gt;        lnL      AIC    AICc Rate.cat ntax</span></span>
<span><span class="co">#&gt;  -177.9706 357.9413 357.951        1  413</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Legend</span></span>
<span><span class="co">#&gt; [1] "colistin_ns"</span></span>
<span><span class="co">#&gt; [1] "0" "1"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Rates</span></span>
<span><span class="co">#&gt;         0       1</span></span>
<span><span class="co">#&gt; 0      NA 89515.8</span></span>
<span><span class="co">#&gt; 1 89515.8      NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Arrived at a reliable solution</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"corHMM_model_summary: A summary of the corHMM model, including the number of parameters, model, number of rate categories, inferred transition rates, log likelihood, AIC, and AICc"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "corHMM_model_summary: A summary of the corHMM model, including the number of parameters, model, number of rate categories, inferred transition rates, log likelihood, AIC, and AICc"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Rate 1 = transitions from level 1 (i.e., susceptible) to level 2 (i.e., non-susceptible)"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Rate 1 = transitions from level 1 (i.e., susceptible) to level 2 (i.e., non-susceptible)"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Rate 2 = transitions from level 2 (i.e., non-susceptible) to level 1 (i.e., susceptible)"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Rate 2 = transitions from level 2 (i.e., non-susceptible) to level 1 (i.e., susceptible)"</span></span>
<span><span class="va">asr_obj</span><span class="op">$</span><span class="va">corHMM_model_summary</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"node_states: Chosen node state that was modeled."</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "node_states: Chosen node state that was modeled."</span></span>
<span><span class="va">asr_obj</span><span class="op">$</span><span class="va">node_states</span></span>
<span><span class="co">#&gt; [1] "joint"</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"parent_child_df: Parent child dataframe, which contains the edge dataset, parent and child values, the child name (for tips), and transition data (i.e., gain, loss, and continuation of the trait on tree or continuation of trait absence on tree)"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "parent_child_df: Parent child dataframe, which contains the edge dataset, parent and child values, the child name (for tips), and transition data (i.e., gain, loss, and continuation of the trait on tree or continuation of trait absence on tree)"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">asr_obj</span><span class="op">$</span><span class="va">parent_child_df</span><span class="op">)</span></span>
<span><span class="co">#&gt;   parent child parent_value child_value child_name transition gain loss</span></span>
<span><span class="co">#&gt; 1    414   415            0           0       &lt;NA&gt;          0    0    0</span></span>
<span><span class="co">#&gt; 2    415   416            0           0       &lt;NA&gt;          0    0    0</span></span>
<span><span class="co">#&gt; 3    416   417            0           0       &lt;NA&gt;          0    0    0</span></span>
<span><span class="co">#&gt; 4    417   418            0           1       &lt;NA&gt;          1    1    0</span></span>
<span><span class="co">#&gt; 5    418   419            1           1       &lt;NA&gt;          0    0    0</span></span>
<span><span class="co">#&gt; 6    419   420            1           1       &lt;NA&gt;          0    0    0</span></span>
<span><span class="co">#&gt;   continuation continuation_present continuation_absent</span></span>
<span><span class="co">#&gt; 1            1                    0                   1</span></span>
<span><span class="co">#&gt; 2            1                    0                   1</span></span>
<span><span class="co">#&gt; 3            1                    0                   1</span></span>
<span><span class="co">#&gt; 4            0                    0                   0</span></span>
<span><span class="co">#&gt; 5            1                    1                   0</span></span>
<span><span class="co">#&gt; 6            1                    1                   0</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-2--synchronous-genotype-phenotype-analyses">Step 2. Synchronous genotype-phenotype analyses<a class="anchor" aria-label="anchor" href="#step-2--synchronous-genotype-phenotype-analyses"></a>
</h2>
<p>For sake of time, we will only run the analysis on the first two
genotypes in this matrix</p>
<pre><code><span><span class="co">#&gt; You specified 'fixed.nodes=FALSE' but included a phy object with node labels. These node labels have been removed.</span></span>
<span><span class="co">#&gt; State distribution in data:</span></span>
<span><span class="co">#&gt; States:  0   1   </span></span>
<span><span class="co">#&gt; Counts:  412 1   </span></span>
<span><span class="co">#&gt; Beginning thorough optimization search -- performing 0 random restarts </span></span>
<span><span class="co">#&gt; Finished. Inferring ancestral states using joint reconstruction. </span></span>
<span><span class="co">#&gt; You specified 'fixed.nodes=FALSE' but included a phy object with node labels. These node labels have been removed.</span></span>
<span><span class="co">#&gt; State distribution in data:</span></span>
<span><span class="co">#&gt; States:  0   1   </span></span>
<span><span class="co">#&gt; Counts:  412 1   </span></span>
<span><span class="co">#&gt; Beginning thorough optimization search -- performing 0 random restarts </span></span>
<span><span class="co">#&gt; Finished. Inferring ancestral states using joint reconstruction.</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Synchronous impact</span></span>
<span><span class="va">synchronous_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">asr_of_genotypes</span>,<span class="va">synchronous_transitions</span>,<span class="va">asr_obj</span><span class="op">$</span><span class="va">parent_child_df</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>,<span class="va">.</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>genotype <span class="op">=</span> <span class="va">variables</span><span class="op">)</span>   </span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Synchronous data frame"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Synchronous data frame"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">synchronous_df</span><span class="op">)</span></span>
<span><span class="co">#&gt;   num_trait_gains synchronous_transitions_num synchronous_gains</span></span>
<span><span class="co">#&gt; 1              37                           0                  </span></span>
<span><span class="co">#&gt; 2              37                           0                  </span></span>
<span><span class="co">#&gt;   synchronous_gains_num synchronous_gains_prop synchronous_gain_loss</span></span>
<span><span class="co">#&gt; 1                     0                      0                      </span></span>
<span><span class="co">#&gt; 2                     0                      0                      </span></span>
<span><span class="co">#&gt;   synchronous_gain_loss_num synchronous_gain_loss_prop num_trait_losses</span></span>
<span><span class="co">#&gt; 1                         0                          0               11</span></span>
<span><span class="co">#&gt; 2                         0                          0               11</span></span>
<span><span class="co">#&gt;   synchronous_losses synchronous_losses_num synchronous_losses_prop</span></span>
<span><span class="co">#&gt; 1                                         0                       0</span></span>
<span><span class="co">#&gt; 2                                         0                       0</span></span>
<span><span class="co">#&gt;   synchronous_loss_gain synchronous_loss_gain_num synchronous_loss_gain_prop</span></span>
<span><span class="co">#&gt; 1                                               0                          0</span></span>
<span><span class="co">#&gt; 2                                               0                          0</span></span>
<span><span class="co">#&gt;                                                                                                                                                                                                                   genotype</span></span>
<span><span class="co">#&gt; 1 Coding SNP at 3359989 &gt; T functional=NULL_NULL_NULL locus_tag=KPNIH1_RS30220 Strand Information: KPNIH1_RS30220=-;T|stop_gained|HIGH|KPNIH1_RS30220|c.141G&gt;A|p.Trp47*|141/144|47/47|mgrB|PhoP/PhoQ regulator MgrB;_68746</span></span>
<span><span class="co">#&gt; 2 Coding SNP at 3360003 &gt; A functional=NULL_NULL_NULL locus_tag=KPNIH1_RS30220 Strand Information: KPNIH1_RS30220=-;A|stop_gained|HIGH|KPNIH1_RS30220|c.127A&gt;T|p.Lys43*|127/144|43/47|mgrB|PhoP/PhoQ regulator MgrB;_68748</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Number of synchronous gain events per genotype"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Number of synchronous gain events per genotype"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">synchronous_df</span><span class="op">$</span><span class="va">synchronous_gains_num</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 0 </span></span>
<span><span class="co">#&gt; 2</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Number of synchronous phenotype gain and genotype loss events per genotype"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Number of synchronous phenotype gain and genotype loss events per genotype"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">synchronous_df</span><span class="op">$</span><span class="va">synchronous_gain_loss_num</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 0 </span></span>
<span><span class="co">#&gt; 2</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Number of synchronous phenotype and genotype loss events per genotype"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Number of synchronous phenotype and genotype loss events per genotype"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">synchronous_df</span><span class="op">$</span><span class="va">synchronous_losses_num</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 0 </span></span>
<span><span class="co">#&gt; 2</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Number of synchronous phenotype los and genotype gain events per genotype"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Number of synchronous phenotype los and genotype gain events per genotype"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">synchronous_df</span><span class="op">$</span><span class="va">synchronous_loss_gain_num</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 0 </span></span>
<span><span class="co">#&gt; 2</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-3-downstream-gainloss-events">Step 3: Downstream gain/loss events<a class="anchor" aria-label="anchor" href="#step-3-downstream-gainloss-events"></a>
</h2>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">downstream_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">asr_of_genotypes</span>,<span class="va">downstream_transitions</span>,<span class="va">asr_obj</span><span class="op">$</span><span class="va">parent_child_df</span>,tr<span class="op">=</span><span class="va">tr</span>,node_states<span class="op">=</span><span class="st">'joint'</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>,<span class="va">.</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>genotype <span class="op">=</span> <span class="va">variables</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Downstream data frame"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Downstream data frame"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">downstream_df</span><span class="op">)</span></span>
<span><span class="co">#&gt;   num_stretches stretches_w_transitions stretches_w_transitions_num</span></span>
<span><span class="co">#&gt; 1             7                    &lt;NA&gt;                           0</span></span>
<span><span class="co">#&gt; 2             7                     647                           1</span></span>
<span><span class="co">#&gt;   stretches_w_gains stretches_w_gains_num stretches_w_gains_prop</span></span>
<span><span class="co">#&gt; 1              &lt;NA&gt;                     0                   0.00</span></span>
<span><span class="co">#&gt; 2               647                     1                   0.14</span></span>
<span><span class="co">#&gt;   stretches_w_losses stretches_w_losses_num stretches_w_losses_prop</span></span>
<span><span class="co">#&gt; 1               &lt;NA&gt;                      0                       0</span></span>
<span><span class="co">#&gt; 2               &lt;NA&gt;                      0                       0</span></span>
<span><span class="co">#&gt;   num_downstream_edges transitions transitions_num transitions_prop gains</span></span>
<span><span class="co">#&gt; 1                  209        &lt;NA&gt;               0      0.000000000      </span></span>
<span><span class="co">#&gt; 2                  209         228               1      0.004784689   228</span></span>
<span><span class="co">#&gt;   gains_num  gains_prop loss loss_num loss_prop</span></span>
<span><span class="co">#&gt; 1         0 0.000000000             0         0</span></span>
<span><span class="co">#&gt; 2         1 0.004784689             0         0</span></span>
<span><span class="co">#&gt;                                                                                                                                                                                                                   genotype</span></span>
<span><span class="co">#&gt; 1 Coding SNP at 3359989 &gt; T functional=NULL_NULL_NULL locus_tag=KPNIH1_RS30220 Strand Information: KPNIH1_RS30220=-;T|stop_gained|HIGH|KPNIH1_RS30220|c.141G&gt;A|p.Trp47*|141/144|47/47|mgrB|PhoP/PhoQ regulator MgrB;_68746</span></span>
<span><span class="co">#&gt; 2 Coding SNP at 3360003 &gt; A functional=NULL_NULL_NULL locus_tag=KPNIH1_RS30220 Strand Information: KPNIH1_RS30220=-;A|stop_gained|HIGH|KPNIH1_RS30220|c.127A&gt;T|p.Lys43*|127/144|43/47|mgrB|PhoP/PhoQ regulator MgrB;_68748</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Number of downstream gain events per genotype"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Number of downstream gain events per genotype"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">downstream_df</span><span class="op">$</span><span class="va">stretches_w_gains_num</span><span class="op">)</span> </span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 0 1 </span></span>
<span><span class="co">#&gt; 1 1</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Number of downstream loss events per genotype"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Number of downstream loss events per genotype"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">downstream_df</span><span class="op">$</span><span class="va">stretches_w_losses_num</span><span class="op">)</span> </span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 0 </span></span>
<span><span class="co">#&gt; 2</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Number of downstream loss events per genotype"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Number of downstream loss events per genotype"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">downstream_df</span><span class="op">$</span><span class="va">stretches_w_losses_num</span><span class="op">)</span> </span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 0 </span></span>
<span><span class="co">#&gt; 2</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Kyle Gontjes.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
