<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>ancestral_state_reconstruction_cluster_detection • phyloAMR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="ancestral_state_reconstruction_cluster_detection">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">phyloAMR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/ancestral_state_reconstruction_cluster_detection.html">ancestral_state_reconstruction_cluster_detection</a></li>
    <li><a class="dropdown-item" href="../articles/ancestral_state_reconstruction_of_a_trait.html">ancestral_state_reconstruction_of_a_trait</a></li>
    <li><a class="dropdown-item" href="../articles/ancestral_state_reconstruction_phenotype_genotype_investigations.html">Phenotype and genotype association testing using ancestral state reconstruction</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/kylegontjes/phyloAMR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>ancestral_state_reconstruction_cluster_detection</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/kylegontjes/phyloAMR/blob/master/vignettes/ancestral_state_reconstruction_cluster_detection.Rmd" class="external-link"><code>vignettes/ancestral_state_reconstruction_cluster_detection.Rmd</code></a></small>
      <div class="d-none name"><code>ancestral_state_reconstruction_cluster_detection.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This tool’s main workhorse is the <strong>asr()</strong>, which
leverages corHMM’s ancestral state reconstruction algorithm to
characterize genome-influenced features’ gain, loss, and continuation
across a phylogenetic tree.</p>
<p>After performing joint or ancestral state reconstruction, the
resultant state predictions of ancestral nodes are parsed to generate a
parent-child dataframe. By traversing the phylogenetic tree from the
tips to the root, the episodes of trait gain, loss, and continuation are
added to the phylogenetic tree’s edge matrix.</p>
<p>This information can be leveraged to infer whether an isolate
belonged to a circulating trait-containing lineage (e.g., evidence of
cross-transmission) or resultant from a within-host emergence event
(e.g., evidence of <em>de novo</em> evolution).</p>
<p>To this aim, we provide a tutorial to leverage our ancestral state
reconstruction wrapper function, asr(), and our tree-traversal
algorithm, asr_cluster_detection(), to infer the evolutionary history of
a trait.</p>
</div>
<div class="section level2">
<h2 id="environment">Environment<a class="anchor" aria-label="anchor" href="#environment"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/kylegontjes/phyloAMR" class="external-link">phyloAMR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/emmanuelparadis/ape" class="external-link">ape</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'ape'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:dplyr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     where</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.amazon.com/Integration-Manipulation-Visualization-Phylogenetic-Computational-ebook/dp/B0B5NLZR1Z/" class="external-link">ggtree</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; ggtree v3.16.0 Learn more at https://yulab-smu.top/contribution-tree-data/</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Please cite:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Guangchuang Yu, David Smith, Huachen Zhu, Yi Guan, Tommy Tsan-Yuk Lam.</span></span>
<span><span class="co">#&gt; ggtree: an R package for visualization and annotation of phylogenetic</span></span>
<span><span class="co">#&gt; trees with their covariates and other associated data. Methods in</span></span>
<span><span class="co">#&gt; Ecology and Evolution. 2017, 8(1):28-36. doi:10.1111/2041-210X.12628</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'ggtree'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:ape':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     rotate</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/johnbaums/hues" class="external-link">hues</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://eliocamp.github.io/ggnewscale/" class="external-link">ggnewscale</a></span><span class="op">)</span> </span></code></pre></div>
</div>
<div class="section level2">
<h2 id="load-example-data">Load example data<a class="anchor" aria-label="anchor" href="#load-example-data"></a>
</h2>
<p>This tutorial will focus on the emergence and spread of colistin
non-susceptibility in a collection of 413 carbapenem-resistant
<em>Klebsiella pneumoniae</em> specimens collected across 12 California
long-term acute care hospitals.</p>
<p>We focus on the evolution and spread of non-susceptibility to
colistin, a last-resort antibiotic used to treat Gram-negative
bacteria.</p>
<div class="section level3">
<h3 id="dataframe">Dataframe<a class="anchor" aria-label="anchor" href="#dataframe"></a>
</h3>
<ol style="list-style-type: decimal">
<li>
<strong>tip_name_variable:</strong> variable with tip names</li>
<li>
<strong>Patient_ID:</strong> Identifiers for patients in this
study</li>
<li>
<strong>clades:</strong> what clade of epidemic lineage sequence
type 258 the isolate belongs to</li>
<li>
<strong>colistin_ns:</strong> colistin non-susceptibility. 1 =
non-susceptible. 0 = susceptible</li>
</ol>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu">phyloAMR</span><span class="fu">::</span><span class="va"><a href="../reference/df.html">df</a></span>  </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Total of 413 isolates"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Total of 413 isolates"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Number of patients: "</span>,<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">Patient_ID</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Number of patients: 338"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 413   4</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"View of dataframe for first 5 isolates"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "View of dataframe for first 5 isolates"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">df</span>,n <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt;         tip_name_var Patient_ID    clades colistin_ns</span></span>
<span><span class="co">#&gt; PCMP_H1      PCMP_H1          1 clade IIA           0</span></span>
<span><span class="co">#&gt; PCMP_H2      PCMP_H2          2 clade IIA           0</span></span>
<span><span class="co">#&gt; PCMP_H3      PCMP_H3          3 clade IIB           1</span></span>
<span><span class="co">#&gt; PCMP_H4      PCMP_H4          4 clade IIB           0</span></span>
<span><span class="co">#&gt; PCMP_H5      PCMP_H5          5   clade I           0</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Frequency of non-susceptibility to colistin: "</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Frequency of non-susceptibility to colistin: "</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">colistin_ns</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   0   1 </span></span>
<span><span class="co">#&gt; 277 136</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="phylogenetic-tree">Phylogenetic tree<a class="anchor" aria-label="anchor" href="#phylogenetic-tree"></a>
</h3>
<p>This maximum-likelihood phylogenetic tree was constructed on a
Gubbins recombination filtered alignment of single-nucleotide
polymorphisms using IQ-TREE</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tr</span> <span class="op">&lt;-</span> <span class="fu">phyloAMR</span><span class="fu">::</span><span class="va"><a href="../reference/tr.html">tr</a></span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html" class="external-link">ggtree</a></span><span class="op">(</span><span class="va">tr</span><span class="op">)</span></span></code></pre></div>
<p><img src="ancestral_state_reconstruction_cluster_detection_files/figure-html/unnamed-chunk-3-1.png" width="720"></p>
</div>
</div>
<div class="section level2">
<h2 id="step-0-visualization-of-the-trait-across-the-phylogenetic-tree">Step 0: Visualization of the trait across the phylogenetic tree<a class="anchor" aria-label="anchor" href="#step-0-visualization-of-the-trait-across-the-phylogenetic-tree"></a>
</h2>
<p>Before performing ancestral state reconstruction, it is critical to
visualize the tip states of the trait on the phylogeny.</p>
<p>Our ancestral state reconstruction and clustering algorithm is most
powerful in settings where frequent emergence and spread of a trait
occurs.</p>
<p>Notice the clustering of Colistin non-susceptibility across the
phylogeny.</p>
<ul>
<li><p>Numerous emergence events at ancestral nodes and tips can be
visually inferred in this phylogeny</p></li>
<li><p>Ancestral emergence events had noticeable variations in size. For
instance, consider the large cluster of non-susceptibility in the above
clade</p></li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>`1` <span class="op">=</span> <span class="st">"black"</span>,`0`<span class="op">=</span><span class="st">"white"</span><span class="op">)</span></span>
<span><span class="va">feature_scale</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="va">feature_colors</span>,labels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Present"</span>,<span class="st">"Absent"</span><span class="op">)</span>,name<span class="op">=</span><span class="st">"Tip State"</span>, guide <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>nrow<span class="op">=</span><span class="fl">1</span>, title.position <span class="op">=</span> <span class="st">"top"</span>, label.position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p0</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/gheatmap.html" class="external-link">gheatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html" class="external-link">ggtree</a></span><span class="op">(</span><span class="va">tr</span><span class="op">)</span>,<span class="va">df</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">colistin_ns</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate_all.html" class="external-link">mutate_all</a></span><span class="op">(</span><span class="va">as.factor</span><span class="op">)</span>,colnames_position <span class="op">=</span> <span class="st">'top'</span>,width <span class="op">=</span> <span class="fl">.25</span>,low <span class="op">=</span> <span class="st">'white'</span>,high<span class="op">=</span><span class="st">'black'</span>,colnames_angle <span class="op">=</span> <span class="fl">90</span>,legend_title <span class="op">=</span> <span class="st">'Colistin non-susceptibility'</span>,hjust <span class="op">=</span> <span class="fl">0</span>,color <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="cn">NA</span>,<span class="fl">450</span><span class="op">)</span> <span class="op">+</span> <span class="va">feature_scale</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">y</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">y</span>, which will replace the existing scale.</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">fill</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span></span>
<span><span class="va">p0</span></span></code></pre></div>
<p><img src="ancestral_state_reconstruction_cluster_detection_files/figure-html/unnamed-chunk-4-1.png" width="720"></p>
</div>
<div class="section level2">
<h2 id="step-1-run-ancestral-state-reconstruction">Step 1: Run ancestral state reconstruction<a class="anchor" aria-label="anchor" href="#step-1-run-ancestral-state-reconstruction"></a>
</h2>
<ul>
<li><p>The workhorse function, asr(), is used to perform ancestral state
reconstruction. This wrapper function implements ancestral state
reconstruction with a single rate category using the corHMM R package:
<a href="https://github.com/thej022214/corHMM" class="external-link uri">https://github.com/thej022214/corHMM</a></p></li>
<li><p>Using inferred ancestral states and tip-based data, edges on the
phylogenetic tree were evaluated to determine episodes where the trait
continued (i.e., susceptible -&gt; susceptible or non-susceptible -&gt;
non-susceptible), was gained (i.e., susceptible -&gt; non-susceptible),
or was lost (i.e., non-susceptible -&gt; susceptible). This edge matrix
can be leveraged for numerous applications, including characterizing the
frequency of trait transitions across a phylogeny and investigating
phenotype-genotype associations.</p></li>
<li>
<p>The following parameters exist for this function</p>
<ul>
<li><p><strong>df:</strong> Dataframe with tip name variable (e.g.,
tip_name_variable) and trait variable (e.g., colistin_ns)</p></li>
<li><p><strong>tr:</strong> Phylogenetic tree object of class
phylo</p></li>
<li>
<p><strong>model:</strong> This approach permits the use of either
the equal rates (ER) or the all rates differ (ARD) transition
matrices.</p>
<ul>
<li><p>Equal rates: Assumes equal transition rates for trait gain (e.g.,
trait absence -&gt; presence) or loss (e.g., trait gain -&gt;
absence)</p></li>
<li><p>All rates differ: Assumes different transition rates for trait
gain and loss</p></li>
</ul>
</li>
<li>
<p>node_states: Whether to perform ‘joint’ or ‘marginal’ ancestral
state reconstruction</p>
<ul>
<li>From our experience, we recommend using joint ancestral state
reconstruction.</li>
</ul>
</li>
</ul>
</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">asr_obj</span> <span class="op">&lt;-</span> <span class="fu">phyloAMR</span><span class="fu">::</span><span class="fu"><a href="../reference/asr.html">asr</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">df</span>,tr <span class="op">=</span> <span class="va">tr</span>,tip_name_variable <span class="op">=</span> <span class="st">"tip_name_var"</span>,trait <span class="op">=</span> <span class="st">"colistin_ns"</span>,model <span class="op">=</span> <span class="st">"ER"</span>,node_states <span class="op">=</span> <span class="st">"joint"</span><span class="op">)</span></span>
<span><span class="co">#&gt; You specified 'fixed.nodes=FALSE' but included a phy object with node labels. These node labels have been removed.</span></span>
<span><span class="co">#&gt; State distribution in data:</span></span>
<span><span class="co">#&gt; States:  0   1   </span></span>
<span><span class="co">#&gt; Counts:  277 136 </span></span>
<span><span class="co">#&gt; Beginning thorough optimization search -- performing 0 random restarts </span></span>
<span><span class="co">#&gt; Finished. Inferring ancestral states using joint reconstruction.</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Output names: "</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Output names: "</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">asr_obj</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "corHMM_output"           "corHMM_model_statistics"</span></span>
<span><span class="co">#&gt; [3] "parent_child_df"         "node_states"</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"corHMM_out: output from ancestral state reconstruction algorithm hosted in the R package corHMM"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "corHMM_out: output from ancestral state reconstruction algorithm hosted in the R package corHMM"</span></span>
<span><span class="va">asr_obj</span><span class="op">$</span><span class="va">corHMM_out</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Fit</span></span>
<span><span class="co">#&gt;        lnL      AIC    AICc Rate.cat ntax</span></span>
<span><span class="co">#&gt;  -177.9706 357.9413 357.951        1  413</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Legend</span></span>
<span><span class="co">#&gt; [1] "colistin_ns"</span></span>
<span><span class="co">#&gt; [1] "0" "1"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Rates</span></span>
<span><span class="co">#&gt;         0       1</span></span>
<span><span class="co">#&gt; 0      NA 89515.8</span></span>
<span><span class="co">#&gt; 1 89515.8      NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Arrived at a reliable solution</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"corHMM_model_summary: A summary of the corHMM model, including the number of parameters, model, number of rate categories, inferred transition rates, log likelihood, AIC, and AICc"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "corHMM_model_summary: A summary of the corHMM model, including the number of parameters, model, number of rate categories, inferred transition rates, log likelihood, AIC, and AICc"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Rate 1 = transitions from level 1 (i.e., susceptible) to level 2 (i.e., non-susceptible)"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Rate 1 = transitions from level 1 (i.e., susceptible) to level 2 (i.e., non-susceptible)"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Rate 2 = transitions from level 2 (i.e., non-susceptible) to level 1 (i.e., susceptible)"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Rate 2 = transitions from level 2 (i.e., non-susceptible) to level 1 (i.e., susceptible)"</span></span>
<span><span class="va">asr_obj</span><span class="op">$</span><span class="va">corHMM_model_summary</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"node_states: Chosen node state that was modeled."</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "node_states: Chosen node state that was modeled."</span></span>
<span><span class="va">asr_obj</span><span class="op">$</span><span class="va">node_states</span></span>
<span><span class="co">#&gt; [1] "joint"</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"parent_child_df: Parent child dataframe, which contains the edge dataset, parent and child values, the child name (for tips), and transition data (i.e., gain, loss, and continuation of the trait on tree or continuation of trait absence on tree)"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "parent_child_df: Parent child dataframe, which contains the edge dataset, parent and child values, the child name (for tips), and transition data (i.e., gain, loss, and continuation of the trait on tree or continuation of trait absence on tree)"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">asr_obj</span><span class="op">$</span><span class="va">parent_child_df</span><span class="op">)</span></span>
<span><span class="co">#&gt;   parent child parent_val child_val child_name transition gain loss</span></span>
<span><span class="co">#&gt; 1    414   415          0         0       &lt;NA&gt;          0    0    0</span></span>
<span><span class="co">#&gt; 2    415   416          0         0       &lt;NA&gt;          0    0    0</span></span>
<span><span class="co">#&gt; 3    416   417          0         0       &lt;NA&gt;          0    0    0</span></span>
<span><span class="co">#&gt; 4    417   418          0         1       &lt;NA&gt;          1    1    0</span></span>
<span><span class="co">#&gt; 5    418   419          1         1       &lt;NA&gt;          0    0    0</span></span>
<span><span class="co">#&gt; 6    419   420          1         1       &lt;NA&gt;          0    0    0</span></span>
<span><span class="co">#&gt;   continuation continuation_present continuation_absent</span></span>
<span><span class="co">#&gt; 1            1                    0                   1</span></span>
<span><span class="co">#&gt; 2            1                    0                   1</span></span>
<span><span class="co">#&gt; 3            1                    0                   1</span></span>
<span><span class="co">#&gt; 4            0                    0                   0</span></span>
<span><span class="co">#&gt; 5            1                    1                   0</span></span>
<span><span class="co">#&gt; 6            1                    1                   0</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-2-characterize-the-transition-statistics-for-the-trait-across-the-phylogeny">Step 2: Characterize the transition statistics for the trait across
the phylogeny<a class="anchor" aria-label="anchor" href="#step-2-characterize-the-transition-statistics-for-the-trait-across-the-phylogeny"></a>
</h2>
<p>The frequency and location of trait gain, loss, and continuation
events can be characterized using the
<strong>asr_transition_analysis()</strong> function</p>
<p>In this case, several important observations occur:</p>
<ol style="list-style-type: decimal">
<li>Of 824 edges, 48 contained transition events: 37 gain events and 11
loss events</li>
<li>Inference of location of these transition events revealed numerous
events at the tips:
<ol style="list-style-type: decimal">
<li>30/37 gain events</li>
<li>7/11 loss events</li>
</ol>
</li>
<li>The number of gain events and large number of non-susceptible
isolates not accounted for by gain events at the tip, suggest potential
for inferred gain events to be shared across isolates.
<ol style="list-style-type: decimal">
<li>This is indicative of the emergence and spread of colistin
non-susceptible strains in this population</li>
</ol>
</li>
<li>We also provide frequency statistics for the gain, loss, and
continuation of these traits that can be useful to characterize the
transition dynamics of this trait:</li>
</ol>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/asr_transition_analysis.html">asr_transition_analysis</a></span><span class="op">(</span><span class="va">asr_obj</span><span class="op">$</span><span class="va">parent_child_df</span>, node_states<span class="op">=</span><span class="st">'joint'</span><span class="op">)</span></span>
<span><span class="co">#&gt;   total_edges transitions gains gains_tip losses losses_tip continuations</span></span>
<span><span class="co">#&gt; 1         824          48    37        30     11          7           776</span></span>
<span><span class="co">#&gt;   continuations_present continuations_absent gain_frequency loss_frequency</span></span>
<span><span class="co">#&gt; 1                   209                  567           6.13              5</span></span>
<span><span class="co">#&gt;   continuation_present_frequency continuation_absent_frequency</span></span>
<span><span class="co">#&gt; 1                          25.36                         68.81</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-3-cluster-detection">Step 3: Cluster detection<a class="anchor" aria-label="anchor" href="#step-3-cluster-detection"></a>
</h2>
<p>Given the numerous gain and loss events, we are well-positioned to
leverage our tree traversal algorithm to infer the evolutionary history
of our isolates.</p>
<p>In this package, we developed a phylogenetic tree traversal
algorithm, <strong>asr_cluster_detection()</strong>, that traces the
ancestral states of each isolate to infer their evolutionary history.
This algorithm takes an isolate’s trait data and walks upward on the
tree to classify trait-containing isolates as phylogenetic singletons
(i.e., evidence of de novo evolution) or members of a phylogenetic
cluster of the trait (i.e., evidence of acquisition of a circulating
trait-containing lineage).</p>
<p>Extending this to the study of Colistin non-susceptibility,
non-susceptible isolates with gain events inferred at the tip were
classified as phylogenetic singletons. However, instances where a
non-susceptible isolate had a gain event at the tip and a reversion
event at its parental node were eligible for classification as members
of a phylogenetic cluster. Non-susceptible isolates were classified as
members of a phylogenetic cluster if their ancestral gain event was
shared with at least one additional resistant isolate. Non-susceptible
isolates that did not share an ancestral gain event with any other
resistant isolate were classified as phylogenetic singletons.
Phylogenetic clusters where all isolates belonged to one patient were
not considered clusters and were reclassified as redundant phylogenetic
singletons. </p>
<p><strong>The following parameters exist for this
function:</strong></p>
<ul>
<li><p><strong>df:</strong> Dataframe with tip name variable (e.g.,
tip_name_variable) and trait variable (e.g., colistin_ns)</p></li>
<li><p><strong>tr:</strong> Phylogenetic tree object of class
phylo</p></li>
<li><p><strong>tip_name_variable:</strong> Tip name variable</p></li>
<li><p><strong>patient_id:</strong> Variable pertaining to a patient
identifier</p></li>
<li><p><strong>pheno:</strong> Phenotype of interest (e.g., colistin
non-susceptibility)</p></li>
<li><p><strong>parent_child_df:</strong> The parent child dataframe
(e.g., edge matrix)</p></li>
<li><p><strong>node_states:</strong> Node states of the ancestral state
reconstruction model</p></li>
<li><p><strong>confidence:</strong> FOR MARGINAL RECONSTRUCTION, ONLY.
Whether to only consider ‘high’ confidence events (e.g., absent to
present) or ‘low’ confidence transition events (e.g., absent to unsure).
We recommend using ‘high’ confidence transition events.</p></li>
<li><p><strong>faux_clusters:</strong> Whether to ‘remove’ (e.g.,
reclassify as singletons) or ‘rename’ (e.g., classify as faux) clusters
where isolates belong to just one patient. This is important to consider
when classifying clusters of this phenotype</p></li>
<li><p><strong>remove_reverant:</strong> Whether to remove (TRUE) the
episodes of trait reversion (e.g., present -&gt; absent) or permit them
to be labeled</p></li>
<li><p><strong>collapse_cluster:</strong> Whether to collapse (TRUE) the
episodes as either cluster, singleton, or no feature</p></li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">asr_cluster_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/asr_cluster_detection.html">asr_cluster_detection</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">df</span>,tr<span class="op">=</span><span class="va">tr</span>,tip_name_variable <span class="op">=</span> <span class="st">"tip_name_var"</span>,patient_id <span class="op">=</span> <span class="st">'PatientID'</span>,parent_child_df <span class="op">=</span> <span class="va">asr_obj</span><span class="op">$</span><span class="va">parent_child_df</span>,node_states <span class="op">=</span> <span class="st">'joint'</span>,confidence <span class="op">=</span> <span class="cn">NULL</span>, simplify_faux_clusters <span class="op">=</span> <span class="cn">FALSE</span>, simplify_revertant <span class="op">=</span> <span class="cn">TRUE</span>, collapse_cluster <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Cluster detection dataframe"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Cluster detection dataframe"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">asr_cluster_obj</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "parent"                "child"                 "parent_val"           </span></span>
<span><span class="co">#&gt;  [4] "child_val"             "child_name"            "transition"           </span></span>
<span><span class="co">#&gt;  [7] "gain"                  "loss"                  "continuation"         </span></span>
<span><span class="co">#&gt; [10] "continuation_present"  "continuation_absent"   "tip_name"             </span></span>
<span><span class="co">#&gt; [13] "asr_cluster"           "asr_cluster_renamed"   "asr_cluster_collapsed"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">asr_cluster_obj</span>,n<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt;           parent child parent_val child_val child_name transition gain loss</span></span>
<span><span class="co">#&gt; PCMP_H325    432     1          1         1  PCMP_H325          0    0    0</span></span>
<span><span class="co">#&gt; PCMP_H411    432     2          1         1  PCMP_H411          0    0    0</span></span>
<span><span class="co">#&gt; PCMP_H213    434     3          1         1  PCMP_H213          0    0    0</span></span>
<span><span class="co">#&gt; PCMP_H286    435     4          1         1  PCMP_H286          0    0    0</span></span>
<span><span class="co">#&gt; PCMP_H387    435     5          1         1  PCMP_H387          0    0    0</span></span>
<span><span class="co">#&gt;           continuation continuation_present continuation_absent  tip_name</span></span>
<span><span class="co">#&gt; PCMP_H325            1                    1                   0 PCMP_H325</span></span>
<span><span class="co">#&gt; PCMP_H411            1                    1                   0 PCMP_H411</span></span>
<span><span class="co">#&gt; PCMP_H213            1                    1                   0 PCMP_H213</span></span>
<span><span class="co">#&gt; PCMP_H286            1                    1                   0 PCMP_H286</span></span>
<span><span class="co">#&gt; PCMP_H387            1                    1                   0 PCMP_H387</span></span>
<span><span class="co">#&gt;           asr_cluster asr_cluster_renamed asr_cluster_collapsed</span></span>
<span><span class="co">#&gt; PCMP_H325 cluster_418           Cluster 1               Cluster</span></span>
<span><span class="co">#&gt; PCMP_H411 cluster_418           Cluster 1               Cluster</span></span>
<span><span class="co">#&gt; PCMP_H213 cluster_418           Cluster 1               Cluster</span></span>
<span><span class="co">#&gt; PCMP_H286 cluster_418           Cluster 1               Cluster</span></span>
<span><span class="co">#&gt; PCMP_H387 cluster_418           Cluster 1               Cluster</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"asr_cluster: raw clustering data. The numbers for cluster and revertant lineages correspond to the ancestral node where the shared gain (i.e., cluster_527) or loss (e.g., revertant_782) event occured."</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "asr_cluster: raw clustering data. The numbers for cluster and revertant lineages correspond to the ancestral node where the shared gain (i.e., cluster_527) or loss (e.g., revertant_782) event occured."</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">asr_cluster_obj</span><span class="op">$</span><span class="va">asr_cluster</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;           cluster_418           cluster_527           cluster_533 </span></span>
<span><span class="co">#&gt;                    72                     2                    12 </span></span>
<span><span class="co">#&gt;           cluster_551           cluster_574           cluster_647 </span></span>
<span><span class="co">#&gt;                     2                     8                     4 </span></span>
<span><span class="co">#&gt;           cluster_775            no feature revertant_cluster_464 </span></span>
<span><span class="co">#&gt;                     6                   236                     2 </span></span>
<span><span class="co">#&gt; revertant_cluster_476 revertant_cluster_583 revertant_cluster_782 </span></span>
<span><span class="co">#&gt;                    15                    15                     2 </span></span>
<span><span class="co">#&gt;         revertant_tip             singleton </span></span>
<span><span class="co">#&gt;                     7                    30</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"asr_cluster_renamed: Renamed these categories as singleton, cluster X, and no feature"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "asr_cluster_renamed: Renamed these categories as singleton, cluster X, and no feature"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">asr_cluster_obj</span><span class="op">$</span><span class="va">asr_cluster_renamed</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Cluster 1  Cluster 2  Cluster 3  Cluster 4  Cluster 5  Cluster 6  Cluster 7 </span></span>
<span><span class="co">#&gt;         72          2         12          2          4          8          6 </span></span>
<span><span class="co">#&gt; No feature  Singleton </span></span>
<span><span class="co">#&gt;        277         30</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"asr_cluster_collapsed: Collapsed as singleton, cluster, or no feature"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "asr_cluster_collapsed: Collapsed as singleton, cluster, or no feature"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">asr_cluster_obj</span><span class="op">$</span><span class="va">asr_cluster_collapsed</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    Cluster No feature  Singleton </span></span>
<span><span class="co">#&gt;        106        277         30</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-4-cluster-statistics">Step 4: Cluster statistics<a class="anchor" aria-label="anchor" href="#step-4-cluster-statistics"></a>
</h2>
<p>To characterize the evolutionary history, we described both the
transitional data and the phylogenetic clustering of each antibiotic
phenotype.</p>
<p>Descriptive statistics for the phylogenetic clustering can be
determined using phyloAMR’s <strong>asr_cluster_analysis()</strong>
function.</p>
<p>Specifically, the crude frequency of the trait, descriptive
statistics on the number of singletons, clusters, and summary statistics
for cluster size.</p>
<p>We use two measures to characterize the phylogenetics of a trait:
phylogenetic occurrence (<strong>Equation 1</strong>) and clustering
(<strong>Equation</strong> <strong>2</strong>). </p>
<ol style="list-style-type: decimal">
<li>
<strong>Phylogenetic frequency:</strong> No. singleton events +
No. clustersNo. singleton events + No. clusters + No. isolates without
feature</li>
<li>
<strong>Clustering frequency:</strong> No. clustersNo. singleton
events + No. clusters  </li>
</ol>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/asr_cluster_analysis.html">asr_cluster_analysis</a></span><span class="op">(</span><span class="va">asr_cluster_obj</span><span class="op">)</span></span>
<span><span class="co">#&gt;   present singletons singleton_isolates clusters cluster_isolates</span></span>
<span><span class="co">#&gt; 1     136         30                 30        7              106</span></span>
<span><span class="co">#&gt;   cluster_size_median cluster_size_mean cluster_size_range no_feature</span></span>
<span><span class="co">#&gt; 1                   6             15.14               2-72        277</span></span>
<span><span class="co">#&gt;   revertant_isolates revertant_clusters revertant_cluster_size_median</span></span>
<span><span class="co">#&gt; 1                 41                  4                           8.5</span></span>
<span><span class="co">#&gt;   revertant_cluster_size_mean revertant_cluster_size_range phylogenetic_events</span></span>
<span><span class="co">#&gt; 1                         8.5                         2-15                  37</span></span>
<span><span class="co">#&gt;   feature_frequency phylogenetic_frequency clustering_frequency</span></span>
<span><span class="co">#&gt; 1             32.93                   7.41                18.92</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-5-visualize-clusters-on-phylogeny">Step 5: Visualize clusters on phylogeny<a class="anchor" aria-label="anchor" href="#step-5-visualize-clusters-on-phylogeny"></a>
</h2>
<p>Finally, it is critical to inspect the lineages on the phylogeny.</p>
<p>Here is a tour of the phylogeny:</p>
<ul>
<li><p>White indicates susceptibility</p></li>
<li>
<p>Red indicates non-susceptible isolate where gain was inferred at
tip</p>
<ul>
<li>Faux clusters (e.g., phylogenetic clusters from one patient) are
also relabeled as red</li>
</ul>
</li>
<li><p>Other colors correspond to the seven clusters of Colistin
non-susceptibility present in this population</p></li>
</ul>
<p>This visualization permits the inspection of cluster calls and
characterize the phylogenetic clustering of our trait using ancestral
state reconstruction</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Cluster color pallete</span></span>
<span><span class="co">## Number of clusters</span></span>
<span><span class="va">ncluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">asr_cluster_obj</span><span class="op">$</span><span class="va">asr_cluster</span><span class="op">)</span>  <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span> <span class="op">!=</span> <span class="st">"singleton"</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"cluster"</span>,<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="va">as.numeric</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span><span class="va">length</span> </span>
<span><span class="co">## Cluster pallete</span></span>
<span><span class="va">color_palette</span> <span class="op">&lt;-</span> <span class="fu">grDevices</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grDevices/colors.html" class="external-link">colors</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"red"</span>,<span class="va">.</span><span class="op">)</span><span class="op">==</span><span class="cn">F</span><span class="op">)</span></span>
<span><span class="va">clusters_col</span> <span class="op">&lt;-</span> <span class="fu">hues</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/hues/man/iwanthue.html" class="external-link">iwanthue</a></span><span class="op">(</span><span class="va">ncluster</span>,hmin<span class="op">=</span><span class="fl">15</span>,hmax<span class="op">=</span><span class="fl">360</span>,lmin <span class="op">=</span> <span class="fl">5</span>,lmax <span class="op">=</span> <span class="fl">95</span>,cmin<span class="op">=</span><span class="fl">5</span>,cmax<span class="op">=</span><span class="fl">90</span>,random <span class="op">=</span> <span class="cn">F</span>,plot <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="co">## Names</span></span>
<span><span class="va">clusters_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Cluster "</span>,<span class="fl">1</span><span class="op">:</span><span class="va">ncluster</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">clusters_col</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">clusters_name</span></span>
<span><span class="co">## Scale</span></span>
<span><span class="va">cluster_scale</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>breaks <span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"No Feature"</span>,<span class="st">"Singleton"</span>,<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">clusters_col</span><span class="op">)</span><span class="op">)</span>,values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"No Feature"</span> <span class="op">=</span> <span class="st">"white"</span>,<span class="st">"Singleton"</span> <span class="op">=</span> <span class="st">"red"</span>,<span class="va">clusters_col</span><span class="op">)</span>,labels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"No Feature"</span>,<span class="st">"NS Singleton"</span>,<span class="va">clusters_name</span><span class="op">)</span>,name<span class="op">=</span><span class="st">"Phylogenetics of Non-susceptibility"</span>, guide <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>ncol<span class="op">=</span><span class="fl">4</span>, title.position <span class="op">=</span> <span class="st">"top"</span>, label.position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Phylogenetic visualization</span></span>
<span><span class="va">p0.1</span> <span class="op">&lt;-</span> <span class="va">p0</span><span class="op">+</span><span class="fu">ggnewscale</span><span class="fu">::</span><span class="fu"><a href="https://eliocamp.github.io/ggnewscale/reference/new_scale.html" class="external-link">new_scale_fill</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/gheatmap.html" class="external-link">gheatmap</a></span><span class="op">(</span><span class="va">p0.1</span>,<span class="va">asr_cluster_obj</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">asr_cluster_renamed</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate_all.html" class="external-link">mutate_all</a></span><span class="op">(</span><span class="va">as.factor</span><span class="op">)</span>,colnames_position <span class="op">=</span> <span class="st">'top'</span>,width <span class="op">=</span> <span class="fl">.25</span>,colnames_angle <span class="op">=</span> <span class="fl">90</span>,legend_title <span class="op">=</span> <span class="st">'Colistin non-usceptibility'</span>,hjust <span class="op">=</span> <span class="fl">0</span>,offset<span class="op">=</span><span class="fl">0.0000075</span>,color<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="cn">NA</span>,<span class="fl">500</span><span class="op">)</span> <span class="op">+</span> <span class="va">cluster_scale</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">y</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">y</span>, which will replace the existing scale.</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">fill</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span></span></code></pre></div>
<p><img src="ancestral_state_reconstruction_cluster_detection_files/figure-html/unnamed-chunk-9-1.png" width="720"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Kyle Gontjes.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer>
</div>





  </body>
</html>
